include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,427 @@
+
+/**
+ * @brief This is a Sparse Format enumeration representing bitflags
+ * 
+ * 
+ */
+enum class SparseFormat : uint32_t {
+  kUndefined = 0x0U,        // For completeness
+  kCoo = 0x1U,              // 1-D or 2-D indices
+  kCsrc = 0x1U << 1,        // Both CSR(C)
+  kBlockSparse = 0x1U << 2  // as in OpenAI
+};
+
+std::ostream& operator<<(std::ostream&, SparseFormat);
+
 /**
  * @brief This class implements SparseTensor. 
- * We represent a SparseTensor as a triple <values, indices, shape>. "values" and "indices" themselves
- * are implemented as Tensors. 
- * We follow the Tensor design for memory ownership/management: a sparse-tensor does not own the "value"
- * or "indices" tensors.
+ * We represent a SparseTensor as a single contiguous buffer with Tensors that project
+ * values into that buffer with other Tensors projecting format specific indices.
+ * 
  */
 
 class SparseTensor final {
  public:
+  /// <summary>
+  /// This constructs an instances that points to user defined buffers.
+  /// Make use of Use* functions to supplied format specific indices that
+  /// reside in the user supplied buffers. Format specific indices are expected
+  /// to be supplied via Use*() methods.
+  /// </summary>
+  /// <param name="elt_type">MlDataType</param>
+  /// <param name="dense_shape">a shape of original tensor in dense form</param>
+  /// <param name="values_shape">shape for user supplied values</param>
+  /// <param name="values_data">a pointer to values</param>
+  /// <param name="location">description of the user allocated memory</param>
   SparseTensor(MLDataType elt_type,
-               const TensorShape& shape,
-               size_t nnz,
+               const TensorShape& dense_shape,
+               const TensorShape& values_shape,
                void* values_data,
-               void* indices_data,
-               const OrtMemoryInfo& memory_info);
+               const OrtMemoryInfo& location);
 
+  /// <summary>
+  /// A constructor that would memorize allocation details but
+  /// will do nothing. All the data would come via Make*() interfaces.
+  /// </summary>
+  /// <param name="elt_type"></param>
+  /// <param name="dense_shape"></param>
+  /// <param name="allocator"></param>
   SparseTensor(MLDataType elt_type,
-               const TensorShape& shape,
-               size_t nnz,
+               const TensorShape& dense_shape,
                std::shared_ptr<IAllocator> allocator);
 
-  ~SparseTensor() = default;
+  SparseTensor() noexcept;
+
+  ~SparseTensor();
 
   // For now, disallow all copy, assignment, and move.
-  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(SparseTensor);
+  ORT_DISALLOW_COPY_AND_ASSIGNMENT(SparseTensor);
 
   // Returns the number of entries in the values tensor (aka "NNZ" or "number of nonzero values")
+  // For block sparse formats this may include some zeros in the blocks are considered non-zero.
   size_t NumValues() const { return static_cast<size_t>(values_.Shape().Size()); }
```

Comment:
Can we cache this at instance creation ? Otherwise every call to NumValues() will re-compute the size.

Reply:
I have to make it mutable

Reply:
I am not seeing a clear case for it as Tensor gets by with this just fine.